src/components/legend/draw.js

```diff
@@ -184,27 +184,51 @@ module.exports = function draw(gd) {
     if(anchorUtils.isRightAnchor(opts)) {
         lx -= opts.width;
     }
-    if(anchorUtils.isCenterAnchor(opts)) {
+    else if(anchorUtils.isCenterAnchor(opts)) {
         lx -= opts.width / 2;
     }
 
     if(anchorUtils.isBottomAnchor(opts)) {
         ly -= opts.height;
     }
-    if(anchorUtils.isMiddleAnchor(opts)) {
+    else if(anchorUtils.isMiddleAnchor(opts)) {
         ly -= opts.height / 2;
     }
 
+    //lx = Math.round(lx);
+    //ly = Math.round(ly);
+
+    // Make sure the legend top and bottom are visible
+    // (legends with a scroll bar are not allowed to stretch beyond the extended
+    // margins)
+    var lyMin = 0;
+    var lyMax = fullLayout.margin.t + fullLayout.height + fullLayout.margin.b;
+    var legendHeight = opts.height;
+    var legendHeightMax = gs.h;
```

Comment:
I'm using the variable names `lyMin`, `lyMax`, `legendHeight` and `legendHeightMax` for documentation purposes (implicitly, to describe the algorithm to compute `legendHeight`).